tests/todini_example_test.cc

```diff
@@ -46,75 +47,155 @@ TEST_CASE("Example network is tested", "[NR method]") {
-                << (*jac) << std::endl
-                << std::endl
-                << "residual = " << std::endl
-                << (*residual_vec) << std::endl
-                << std::endl
-                << "variable = " << std::endl
-                << (*variable_vec) << std::endl
-                << std::endl;
-      break;
-    }
-    std::cout << "Jac = " << std::endl
-              << (*jac) << std::endl
-              << std::endl
-              << "residual = " << std::endl
-              << (*residual_vec) << std::endl
-              << std::endl
-              << "variable = " << std::endl
-              << (*variable_vec) << std::endl
-              << std::endl;
-
-    nr_iter++;
+  mesh->assign_node_elevation (init_nodal_head);
+  mesh->assign_node_demand (init_nodal_discharge);
+
+  SECTION ("DD mode test"){
+      // Initialize matrix assembler and obtain global index to nodes and pipes
+      bool pdd_mode = false;
+      auto assembler = std::make_shared<pipenetwork::MatrixAssembler>(pdd_mode);
+      assembler->global_nodal_pipe_indices(mesh);
+
+      // Assemble variable vector, residual vector and Jacobian
+      assembler->assemble_variable_vector();
+      std::shared_ptr<Eigen::VectorXd> variable_vec = assembler->variable_vec();
+      assembler->assemble_residual_vector();
+      std::shared_ptr<Eigen::VectorXd> residual_vec = assembler->residual_vec();
+      assembler->assemble_jacobian();
+      std::shared_ptr<Eigen::SparseMatrix<double>> jac = assembler->jac();
+
+      // Creat a eigen gmres solver and solve
+      double gmres_tolerance = 1.e-12;
+      const unsigned max_iter = 5000;
+      auto solver =
+              std::make_shared<pipenetwork::EigenGMRES>(max_iter, gmres_tolerance);
+      solver->assembled_matrices(jac, variable_vec, residual_vec);
+
+      //  // Apply restraints
+      //  Eigen::VectorXd restraints(17);
+      //  restraints << 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1;
+      //  solver->restrains(restraints);
+//        std::cout << "Jac = " << std::endl
+//                  << (*jac) << std::endl
+//                  << std::endl
+//                  << "residual = " << std::endl
+//                  << std::endl
+//                  << (*residual_vec) << std::endl
+//                  << "variable = " << std::endl
+//                  << (*variable_vec) << std::endl
+//                  << std::endl;
+      bool issolved = solver->solve();
+      assembler->apply_variables();
+      assembler->assemble_residual_vector();
+//        std::cout << "Jac = " << std::endl
+//                  << (*jac) << std::endl
+//                  << std::endl
+//                  << "residual = " << std::endl
+//                  << std::endl
+//                  << (*residual_vec) << std::endl
+//                  << "variable = " << std::endl
+//                  << (*variable_vec) << std::endl
+//                  << std::endl;
+
+      unsigned nr_iter = 1;
+      while (nr_iter < 30) {
```

Comment:
Does it converge in 30 iterations?

Reply:
why 30?

Reply:
It is just an arbitrary number for testing purpose. Usually it converges in 10+.